src/librbd/ImageWatcher.cc

```diff
@@ -67,7 +67,10 @@ ImageWatcher::~ImageWatcher()
 
 bool ImageWatcher::is_lock_supported() const {
   assert(m_image_ctx.owner_lock.is_locked());
-  return ((m_image_ctx.features & RBD_FEATURE_EXCLUSIVE_LOCK) != 0 &&
+  RWLock::RLocker l(m_image_ctx.snap_lock);
+  uint64_t snap_features;
+  m_image_ctx.get_features(m_image_ctx.snap_id, &snap_features);
+  return ((snap_features & RBD_FEATURE_EXCLUSIVE_LOCK) != 0 &&
```

Comment:
Minor: exclusive lock can only be acquired on HEAD, so you could have tested m_image_ctx.features while holding snap_lock.